File: common/src/main/java/juuxel/adorn/block/entity/KitchenSinkBlockEntity.java
package juuxel.adorn.block.entity;

import juuxel.adorn.block.AdornBlockEntities;
import juuxel.adorn.fluid.FluidReference;
import juuxel.adorn.lib.AdornGameRules;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.core.component.DataComponents;
import net.minecraft.world.item.alchemy.PotionContents;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.material.FlowingFluid;
import net.minecraft.world.level.material.Fluid;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.item.alchemy.Potions;
import net.minecraft.core.HolderLookup;
import net.minecraft.tags.FluidTags;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundSource;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.world.InteractionHand;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.gameevent.GameEvent;
import org.jetbrains.annotations.Nullable;

public abstract class KitchenSinkBlockEntity extends BlockEntity {
    public KitchenSinkBlockEntity(BlockPos pos, BlockState state) {
        super(AdornBlockEntities.KITCHEN_SINK.get(), pos, state);
    }

    /**
     * A reference to the current fluid contents of this sink.
     */
    public abstract FluidReference getFluidReference();

    /**
     * Tries to interact with this kitchen sink with a fluid container stack.
     * @return true if inserted and false otherwise
     */
    public abstract boolean interactWithItem(ItemStack stack, Player player, InteractionHand hand);

    /**
     * Clears all fluids from this kitchen sink.
     * @return true if cleared and false otherwise
     */
    public abstract boolean clearFluidsWithSponge();

    /**
     * Called when this kitchen sink is filled with fluids.
     * Dispatches the game event and plays the sound.
     */
    protected void onFill(ItemStack stack, Player player) {
        level.gameEvent(player, GameEvent.FLUID_PLACE, worldPosition);
        player.playSound(getEmptySound(getFluidReference(), stack).event, 1f, 1f);
    }

    /**
     * Called when fluids are picked up from this kitchen sink.
     * Dispatches the game event and plays the sound.
     */
    protected void onPickUp(FluidReference fluid, ItemStack stack, Player player) {
        level.gameEvent(player, GameEvent.FLUID_PICKUP, worldPosition);
        player.level().playSound(player, player.getX(), player.getY(), player.getZ(), getFillSound(fluid, stack).event, SoundSource.BLOCKS, 1f, 1f);
    }

    protected FluidItemSound getFillSound(FluidReference fluid, ItemStack stack) {
        if (stack.is(Items.GLASS_BOTTLE)) {
            return new FluidItemSound(SoundEvents.BOTTLE_FILL, true);
        }

        return new FluidItemSound(fluid.getFluid().getPickupSound().orElse(SoundEvents.BUCKET_FILL), false);
    }

    protected FluidItemSound getEmptySound(FluidReference fluid, ItemStack stack) {
        if (isWaterBottle(stack)) {
            return new FluidItemSound(SoundEvents.BOTTLE_EMPTY, true);
        }

        return new FluidItemSound(fluid.getFluid().is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY, false);
    }

    protected static boolean isWaterBottle(ItemStack stack) {
        if (!stack.is(Items.POTION)) return false;
        var potionContents = stack.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY);
        return potionContents.is(Potions.WATER);
    }

    protected void markDirtyAndSync() {
        setChanged();

        if (!level.isClientSide()) {
            level.sendBlockUpdated(worldPosition, getBlockState(), getBlockState(), Block.UPDATE_ALL);
        }
    }

    @Override
    public Packet<ClientGamePacketListener> getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }

    @Override
    public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
        return saveCustomOnly(registries);
    }

    /**
     * Calculates the comparator output based on tank contents.
     */
    public abstract int calculateComparatorOutput();

    private static boolean isInfinite(Fluid fluid, ServerLevel world) {
        return fluid instanceof FlowingFluid flowable && flowable.canConvertToSource(world);
    }

    public static boolean supportsInfiniteExtraction(ServerLevel world, Fluid fluid) {
        return isInfinite(fluid, world) && world.getGameRules().get(AdornGameRules.INFINITE_KITCHEN_SINKS.get());
    }

    /**
     * A sound event containing coupled with whether it's preferred.
     * Used for a cursed priority system for fill/empty sounds:
     * bottle sounds are preferred by default, then Forge's fluid sounds and finally vanilla sounds.
     */
    public record FluidItemSound(
        SoundEvent event, boolean preferred) {
        public FluidItemSound orElse(@Nullable SoundEvent fallback) {
            if (preferred) return this;
            return fallback != null ? new FluidItemSound(fallback, true) : this;
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/entity/KitchenSinkBlockEntity.kt
package juuxel.adorn.block.entity

import juuxel.adorn.block.AdornBlockEntities
import juuxel.adorn.fluid.FluidReference
import juuxel.adorn.lib.AdornGameRules
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.fluid.FlowableFluid
import net.minecraft.fluid.Fluid
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.nbt.NbtCompound
import net.minecraft.network.Packet
import net.minecraft.network.listener.ClientPlayPacketListener
import net.minecraft.network.packet.s2c.play.BlockEntityUpdateS2CPacket
import net.minecraft.potion.PotionUtil
import net.minecraft.potion.Potions
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvent
import net.minecraft.sound.SoundEvents
import net.minecraft.tag.FluidTags
import net.minecraft.util.Hand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.event.GameEvent

abstract class KitchenSinkBlockEntity(pos: BlockPos, state: BlockState) : BlockEntity(AdornBlockEntities.KITCHEN_SINK, pos, state) {
    /** A reference to the current fluid contents of this sink. */
    abstract val fluidReference: FluidReference

    /**
     * Tries to interact with this kitchen sink with a fluid container [stack].
     * @return true if inserted and false otherwise
     */
    abstract fun interactWithItem(stack: ItemStack, player: PlayerEntity, hand: Hand): Boolean

    /**
     * Clears all fluids from this kitchen sink.
     * @return true if cleared and false otherwise
     */
    abstract fun clearFluidsWithSponge(): Boolean

    /**
     * Called when this kitchen sink is filled with fluids.
     * Dispatches the game event and plays the sound.
     */
    protected fun onFill(stack: ItemStack, player: PlayerEntity) {
        val w = world!!
        if (!w.isClient) {
            w.emitGameEvent(player, GameEvent.FLUID_PLACE, pos)
            player.playSound(getEmptySound(fluidReference, stack).event, SoundCategory.BLOCKS, 1f, 1f)
        }
    }

    /**
     * Called when fluids are picked up from this kitchen sink.
     * Dispatches the game event and plays the sound.
     */
    protected fun onPickUp(fluid: FluidReference, stack: ItemStack, player: PlayerEntity) {
        val w = world!!
        if (!w.isClient) {
            w.emitGameEvent(player, GameEvent.FLUID_PICKUP, pos)
            player.playSound(getFillSound(fluid, stack).event, SoundCategory.BLOCKS, 1f, 1f)
        }
    }

    protected open fun getFillSound(fluid: FluidReference, stack: ItemStack): FluidItemSound {
        if (stack.isOf(Items.GLASS_BOTTLE)) {
            return FluidItemSound(SoundEvents.ITEM_BOTTLE_FILL, true)
        }

        return FluidItemSound(fluid.fluid.bucketFillSound.orElse(SoundEvents.ITEM_BUCKET_FILL), false)
    }

    protected open fun getEmptySound(fluid: FluidReference, stack: ItemStack): FluidItemSound {
        if (stack.isOf(Items.POTION) && PotionUtil.getPotion(stack) == Potions.WATER) {
            return FluidItemSound(SoundEvents.ITEM_BOTTLE_EMPTY, true)
        }

        return FluidItemSound(if (fluid.fluid.isIn(FluidTags.LAVA)) SoundEvents.ITEM_BUCKET_EMPTY_LAVA else SoundEvents.ITEM_BUCKET_EMPTY, false)
    }

    protected fun markDirtyAndSync() {
        markDirty()

        val w = world!!
        if (!w.isClient) {
            w.updateListeners(pos, cachedState, cachedState, 3)
        }
    }

    override fun toUpdatePacket(): Packet<ClientPlayPacketListener> =
        BlockEntityUpdateS2CPacket.create(this)

    override fun toInitialChunkDataNbt(): NbtCompound = createNbt()

    /** Calculates the comparator output based on tank contents. */
    abstract fun calculateComparatorOutput(): Int

    companion object {
        private fun isInfinite(fluid: Fluid): Boolean =
            fluid is FlowableFluid && fluid.isInfinite

        fun supportsInfiniteExtraction(world: World, fluid: Fluid): Boolean =
            isInfinite(fluid) && world.gameRules.getBoolean(AdornGameRules.INFINITE_KITCHEN_SINKS)
    }

    /**
     * A sound event containing coupled with whether it's preferred.
     * Used for a cursed priority system for fill/empty sounds:
     * bottle sounds are preferred by default, then Forge's fluid sounds and finally vanilla sounds.
     */
    data class FluidItemSound(val event: SoundEvent, val preferred: Boolean) {
        fun orElse(fallback: SoundEvent?): FluidItemSound {
            if (preferred) return this
            return if (fallback != null) FluidItemSound(fallback, true) else this
        }
    }
}
